### PR TITLE
Add logic to variation title generation and remove redundant template info.

### PR DESCRIPTION
**Pull request by claudiulodro**
This relates to #12990.

I've implemented @mikejolley's suggestion to use the shortform variation title when a variation only has one attribute and use the longform title when a variation has more than one attribute. In testing, this seems to work great, and I think it has solved the problem.

I have also removed the redundant attribute list under the product name on the order pages, as it's now very clear what attributes a product has by its name.

### CODE REVIEW

File: includes/data-stores/class-wc-product-variation-data-store-cpt.php [hunk 1]
```diff
@@ -165,6 +164,23 @@ public function update( &$product ) {
...
+	 * Generates a title with attribute information for a variation.
+	 * Products with one attribute will get a title of the form "Name - Value"
+	 * Products with 2+ attributes will get a title of the form "Name - Attribute: Value, Attribute: Value"
+	 *
+	 * @since 2.7.0
+	 * @param WC_Product
+	 * @return string
+	 */
+	protected function generate_product_title( $product ) {
+		$use_longform_title = apply_filters( 'woocommerce_product_variation_title_use_longform_style', count( $product->get_attributes() ) > 1, $product );
```
Comment (mikejolley) [inline includes/data-stores/class-wc-product-variation-data-store-cpt.php]: I'm not sure 'use_longform_style' is clear enough. How about describing what we're actually doing? i.e. `woocommerce_product_variation_title_include_attribute_names`. True if we include them here.

File: includes/data-stores/class-wc-product-variation-data-store-cpt.php [hunk 2]
```diff
@@ -102,7 +101,7 @@ public function create( &$product ) {
 			'post_type'      => 'product_variation',
 			'post_status'    => $product->get_status() ? $product->get_status() : 'publish',
 			'post_author'    => get_current_user_id(),
-			'post_title'     => $parent_object->post_title . ' &ndash; ' . wc_get_formatted_variation( $product, true, false ),
```
Comment (mikejolley) [inline includes/data-stores/class-wc-product-variation-data-store-cpt.php]: If `parent_object` is not used in the method, it should be removed too.

File: includes/data-stores/class-wc-product-variation-data-store-cpt.php [hunk 3]
```diff
@@ -165,6 +164,23 @@ public function update( &$product ) {
...
+	 * @since 2.7.0
+	 * @param WC_Product
+	 * @return string
+	 */
+	protected function generate_product_title( $product ) {
+		$use_longform_title = apply_filters( 'woocommerce_product_variation_title_use_longform_style', count( $product->get_attributes() ) > 1, $product );
+		$title_base = apply_filters( 'woocommerce_product_variation_title_base', get_post_field( 'post_title', $product->get_parent_id() ), $product );
+		$separator = apply_filters( 'woocommerce_product_variation_title_separator', '&ndash;', $product );
+
+		return $title_base . ' ' . $separator . ' ' . wc_get_formatted_variation( $product, true, $use_longform_title );
```
Comment (mikejolley) [inline includes/data-stores/class-wc-product-variation-data-store-cpt.php]: Filter the return value and pass in all parts.

File: includes/data-stores/class-wc-product-variation-data-store-cpt.php [hunk 4]
```diff
@@ -165,6 +164,23 @@ public function update( &$product ) {
...
+	 * Products with 2+ attributes will get a title of the form "Name - Attribute: Value, Attribute: Value"
+	 *
+	 * @since 2.7.0
+	 * @param WC_Product
+	 * @return string
+	 */
+	protected function generate_product_title( $product ) {
+		$use_longform_title = apply_filters( 'woocommerce_product_variation_title_use_longform_style', count( $product->get_attributes() ) > 1, $product );
+		$title_base = apply_filters( 'woocommerce_product_variation_title_base', get_post_field( 'post_title', $product->get_parent_id() ), $product );
+		$separator = apply_filters( 'woocommerce_product_variation_title_separator', '&ndash;', $product );
```
Comment (mikejolley) [inline includes/data-stores/class-wc-product-variation-data-store-cpt.php]: I don't think we need a separator filter. If we filter the return value, thats sufficient IMO

File: templates/emails/email-order-items.php [hunk 5]
```diff
@@ -43,8 +43,6 @@
 				// allow other plugins to add additional product information here
 				do_action( 'woocommerce_order_item_meta_start', $item_id, $item, $order, $plain_text );
 
-				wc_display_item_meta( $item );
```
Comment (mikejolley) [inline templates/emails/email-order-items.php]: This needs a different approach. This method outputs other non-attribute data too e.g. if using product addons extension, it will list other selected values which won't be included in the title. We should instead modify this method to skip over attribute values which I believe have a `pa_` prefix.

File: templates/emails/plain/email-order-items.php [hunk 6]
```diff
@@ -31,13 +31,6 @@
 		echo ' = ' . $order->get_formatted_line_subtotal( $item ) . "\n";
 		// allow other plugins to add additional product information here
 		do_action( 'woocommerce_order_item_meta_start', $item_id, $item, $order, $plain_text );
-		echo strip_tags( wc_display_item_meta( $item, array(
```
Comment (mikejolley) [inline templates/emails/plain/email-order-items.php]: See https://github.com/woocommerce/woocommerce/pull/13136/files#r100519293

File: templates/order/order-details-item.php [hunk 7]
```diff
@@ -36,7 +36,6 @@
 
 			do_action( 'woocommerce_order_item_meta_start', $item_id, $item, $order );
 
-			wc_display_item_meta( $item );
```
Comment (mikejolley) [inline templates/order/order-details-item.php]: See https://github.com/woocommerce/woocommerce/pull/13136/files#r100519293

**Review by mikejolley — CHANGES_REQUESTED** [review]
Added some notes :)

I quite liked your idea about the 'attributes with 1 word' being potentially valid for short form too. Do you think it's worth accounting for that in `generate_product_title` too?

**Comment by claudiulodro** [comment]
This was my approach for handling attribute suppression in `wc_display_item_meta`:

1. Expose the beginning 'attribute_' part of the meta key. `WC_Order_Item::get_formatted_meta_data` already accounts for 'attribute_' in the meta key, so it shouldn't break anything.
2. Modify `WC_Order_Item::get_formatted_meta_data` to accept longer hideprefixes.
3. Pass in 'attribute_' as a hideprefix in `wc_display_item_meta`

This should just hide attributes in `wc_display_item_meta` while allowing other meta through.

File: includes/class-wc-order-item-product.php [hunk 8]
```diff
@@ -167,7 +167,7 @@ public function set_taxes( $raw_tax_data ) {
 	 */
 	public function set_variation( $data ) {
 		foreach ( $data as $key => $value ) {
-			$this->add_meta_data( str_replace( 'attribute_', '', $key ), $value, true );
```
Comment (mikejolley) [inline includes/class-wc-order-item-product.php]: This change might be an issue for backwards compatibility. I believe some code uses that prefix to identify attribute rows. Why was this removed can I ask?

Comment (claudiulodro) [inline includes/class-wc-order-item-product.php]: That prefix was getting removed, and I modified it to keep the prefix, as that was the best way to reliably determine which meta values were attributes. For my clarity, does having the prefix or not having the prefix break backwards compatibility?

Comment (mikejolley) [inline includes/class-wc-order-item-product.php]: Now you corrected me I'm not sure :) When they are stored in the backend (meta) do they have a prefix? If thats consistent with 2.6.x it should be fine.

Comment (claudiulodro) [inline includes/class-wc-order-item-product.php]: https://github.com/woocommerce/woocommerce/blob/2.6.14/includes/abstracts/abstract-wc-order.php#L301 It looks like 2.6 removes the prefix. I'll modify this to preserve that same functionality.

**Comment by claudiulodro** [comment]
I've modified `generate_product_title ` to handle titles with no attributes and tweaked the way meta data is filtered in `WC_Order_Item::get_formatted_meta_data `. Instead of modifying the metadata like the previous implementation, there's a flag that just excludes a product's attributes from the output. This should preserve backwards compatibility.

**Comment by mikejolley** [comment]
Titles are fixed now 👍 

Seeing one more issue after ordering. 'test' in this screenshot has an 'any' attribute selected which showed fine in cart, but after ordering the data is missing:

2017-02-14 at 12 12

2017-02-14 at 12 13

Seems to be the `include_attributes` change. I wonder if we can do this conditionally on output...hopefully without having to pull in too much product data to make it possible. i.e. if the meta key appears to be an attribute (they have a prefix?) and if the value appears to be in the product title, skip onto the next item of meta.

Thoughts?

**Comment by claudiulodro** [comment]
Good call. I wasn't able to keep the prefix on the meta for backwards compatibility. If the any-attribute-variation is the only case we have to look out for, I can check if the title is the same as the parent title, and output the list if it is. That should solve the issue.

**Comment by claudiulodro** [comment]
I believe I've got it worked out. Just skipping attributes with values in the product's name in `get_formatted_meta_data` should work great. The resulting code is way simpler and I believe it should work for all reasonable situations.

screen shot 2017-02-14 at 11 26 29 am

**Review by mikejolley — APPROVED** [review]
LGTM 👍

**Review by claudiosanches — APPROVED** [review]
Looks good!
